stdlib/LinearAlgebra/src/LinearAlgebra.jl

```diff
@@ -479,22 +479,26 @@ wrapper_char(A::Hermitian) = A.uplo == 'U' ? 'H' : 'h'
 wrapper_char(A::Hermitian{<:Real}) = A.uplo == 'U' ? 'S' : 's'
 wrapper_char(A::Symmetric) = A.uplo == 'U' ? 'S' : 's'
 
-Base.@constprop :aggressive function wrap(A::AbstractVecOrMat, tA::AbstractChar)
-    if tA == 'N'
-        return A
+Base.@constprop :aggressive function wrap(A::AbstractVecOrMat{T}, tA::AbstractChar) where {T}
+    # merge the result of this before return, so that we can type-assert the return such
+    # that even if the tmerge is inaccurate, inference can still recover precise info on the
+    # eltype and ndims and avoid excess backedges in the consumers of this info (e.g. _generic_matmatmul!)
+    B = if tA == 'N'
+        A
     elseif tA == 'T'
-        return transpose(A)
+        transpose(A)
     elseif tA == 'C'
-        return adjoint(A)
+        adjoint(A)
     elseif tA == 'H'
-        return Hermitian(A, :U)
+        Hermitian(A, :U)
     elseif tA == 'h'
-        return Hermitian(A, :L)
+        Hermitian(A, :L)
     elseif tA == 'S'
-        return Symmetric(A, :U)
+        Symmetric(A, :U)
     else # tA == 's'
-        return Symmetric(A, :L)
+        Symmetric(A, :L)
     end
+    return A isa AbstractVector ? B::AbstractVector{T} : B::AbstractMatrix{T}
```

Comment:
I think the main question to confirm is if this type assert is correct for eltype and ndims. If not, we can relax it to something more generic such as just `AbstractVecOrMat`, but we lose some accuracy perhaps.

Reply:
Alright, looks like CI has counter examples to both ndims and eltype, such as

```
LinearAlgebra.Transpose{LinearAlgebra.Transpose{ComplexF64, Vector{ComplexF64}}, Matrix{Vector{ComplexF64}}}
```

```suggestion
    return B::AbstractVecOrMat
```

Reply:
I'm unsure if this helps, but the `ndims` assert should work the other way: if `A isa AbstractMatrix`, then so is `B`